File: lib_audio/src/main/java/com/imooc/lib_audio/mediaplayer/core/AudioController.java
package com.imooc.lib_audio.mediaplayer.core;

import android.util.Log;

import com.imooc.lib_audio.app.AudioHelper;
import com.imooc.lib_audio.mediaplayer.events.AudioCompleteEvent;
import com.imooc.lib_audio.mediaplayer.events.AudioErrorEvent;
import com.imooc.lib_audio.mediaplayer.events.AudioPlayModeEvent;
import com.imooc.lib_audio.mediaplayer.events.AudioRemoveEvent;
import com.imooc.lib_audio.mediaplayer.exception.AudioQueueEmptyException;
import com.imooc.lib_api.model.AudioBean;
import com.imooc.lib_common_ui.utils.SharePreferenceUtil;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

//不允许在包外的类访问
public class AudioController {

	private static final String TAG = "AudioController";

	/*
	 * 播放模式
	 */
	public enum PlayMode{
		LOOP,
		RANDOM,
		REPEAT
	}

	//播放器
	private final AudioPlayer mAudioPlayer;
	//当前播放队列
	private final ArrayList<AudioBean> mQueue = new ArrayList<>();
	//播放模式
	private PlayMode mPlayMode;
	//当前播放索引
	private int mQueueIndex;

	private AudioController(){
		mAudioPlayer = new AudioPlayer();
		final AudioBean audio = SharePreferenceUtil.getInstance(AudioHelper.getContext()).getLatestSong();
		final List<AudioBean> list = SharePreferenceUtil.getInstance(AudioHelper.getContext()).getMusicList();
		if(audio != null && list != null) {
			mQueue.add(audio);
			list.remove(audio);
			//上次保存的播放列表
			mQueue.addAll(list);
		}
		mQueueIndex = 0;
		mPlayMode = PlayMode.LOOP;
		EventBus.getDefault().register(this);

	}

	public static AudioController getInstance(){
		return SingletonHolder.INSTANCE;
	}

	private static class SingletonHolder{
		private static final AudioController INSTANCE = new AudioController();
	}

	public boolean isStartState() {
		return CustomMediaPlayer.Status.STARTED == getStatus();
	}

	/**
	 * 对外提提供是否暂停状态
	 */
	private boolean isPauseState() {
		return CustomMediaPlayer.Status.PAUSED == getStatus();
	}

	private CustomMediaPlayer.Status getStatus() {
		return mAudioPlayer.getState();
	}

	public ArrayList<AudioBean> getQueue() {
		return mQueue;
	}

	//添加Audio
	public void addAudio(AudioBean bean) {
		addAudio(0, bean);
	}

	public void addAudio(ArrayList<AudioBean> list) {
		mQueue.addAll(list);
	}

	public void addAudio(int index, AudioBean bean) {
		if(mQueue.size() <= 0){
			throw new AudioQueueEmptyException("");
		}
		int query = mQueue.indexOf(bean);
		if(query <= -1){
			//当前播放列表中没有此歌曲
			mQueue.add(index, bean);
			setPlayIndex(index);
		}else{
			//当前播放列表中有这个歌曲
			AudioBean currentPlaying = getCurrentPlaying();
			if(!currentPlaying.getId().equals(bean.getId())){
				setPlayIndex(query);
			}
		}
	}

	//删除单个Audio
	public void removeAudio(AudioBean bean){
		if(mQueue.size() <= 0){
			throw new AudioQueueEmptyException("");
		}
		if(mQueue.remove(bean)){
			EventBus.getDefault().post(new AudioRemoveEvent());
		}
	}

	//删除全部
	public void removeAudio(){
		if(mQueue.size() <= 0){
			throw new AudioQueueEmptyException("");
		}
		AudioBean currentBean = getCurrentPlaying();
		mQueue.clear();
		//不删除当前播放的歌曲
		mQueue.add(currentBean);

	}

	public void setQueue(ArrayList<AudioBean> bean){
		setQueue(bean, 0);
	}

	public void setQueue(ArrayList<AudioBean> bean, int index){
		mQueue.addAll(bean);
		mQueueIndex = index;
	}

	public PlayMode getPlayMode() {
		return mPlayMode;
	}

	public void setPlayMode(PlayMode playMode) {
		mPlayMode = playMode;
		//还要对外发送切换事件，更新UI
		EventBus.getDefault().post(new AudioPlayModeEvent(mPlayMode));
		Log.e(TAG, "AudioPlayModeEvent " + mPlayMode.name());
	}

	public void setPlayIndex(int index) {
		if (mQueue.size() <= 0) {
			throw new AudioQueueEmptyException("当前播放队列为空,请先设置播放队列.");
		}
		mQueueIndex = index;
		play();
	}

	public void play() {
		mAudioPlayer.load(getCurrentPlaying());
	}

	private AudioBean getCurrentPlaying() {
		return getPlaying(mQueueIndex);
	}

	public void pause(){
		mAudioPlayer.pause();
	}

	public void seekTo(long time){
		mAudioPlayer.seekTo(time);
	}
	public void release(){
		mAudioPlayer.release();
		EventBus.getDefault().unregister(this);
	}

	public void resume(){
		mAudioPlayer.resume();
	}

	public void next(){
		mAudioPlayer.load(getNextPlaying());
	}

	//获取下一首歌曲
	private AudioBean getNextPlaying() {
		switch (mPlayMode){
			case LOOP:
				//列表循环 取余运算防止越界
				mQueueIndex = (mQueueIndex + 1) % mQueue.size();
				return getPlaying(mQueueIndex);
			case RANDOM:
				//随机播放
				mQueueIndex = new Random().nextInt(mQueue.size()) % mQueue.size();
				return getPlaying(mQueueIndex);
			case REPEAT:
				//单曲循环
				return getPlaying(mQueueIndex);
			default:
				break;
		}
		return null;
	}

	//获取上一首歌曲
	private AudioBean getPrePlaying() {
		switch (mPlayMode) {
			case LOOP:
				mQueueIndex = (mQueueIndex + mQueue.size() - 1) % mQueue.size();
				return getPlaying(mQueueIndex);
			case RANDOM:
				mQueueIndex = new Random().nextInt(mQueue.size()) % mQueue.size();
				return getPlaying(mQueueIndex);
			case REPEAT:
				return getPlaying(mQueueIndex);
		}
		return null;
	}

	public void previous(){
		mAudioPlayer.load(getPrePlaying());
	}

	public int getQueueIndex() {
		return mQueueIndex;
	}

	/**
	 * 切换播放/暂停状态
	 */
	public void playOrPause(){
		if(CustomMediaPlayer.Status.IDLE == getStatus()){
			play();
		}
		if(isStartState()){
			pause();
		}else if(isPauseState()){
			resume();
		}
	}

	private AudioBean getPlaying(int index) {
		if (mQueue != null && !mQueue.isEmpty() && index >= 0 && index < mQueue.size()) {
			return mQueue.get(index);
		} else {
			return null;
			//throw new AudioQueueEmptyException("当前播放队列为空,请先设置播放队列." + "当前Index: "+ index + "播放队列大小: " + mQueue.size());
		}
	}

	public AudioBean getNowPlaying() {
		return getPlaying(mQueueIndex);
	}

	//插放完毕事件处理
	@Subscribe(threadMode = ThreadMode.MAIN)
	public void onAudioCompleteEvent(AudioCompleteEvent event) {
		Log.e(TAG, "onAudioCompleteEvent");
		next();
	}

	//播放出错事件处理
	@Subscribe(threadMode = ThreadMode.MAIN)
	public void onAudioErrorEvent(AudioErrorEvent event) {
		next();
	}

}

File: lib_audio/src/main/java/com/imooc/lib_audio/mediaplayer/core/AudioController.kt
package com.imooc.lib_audio.mediaplayer.core

import com.imooc.lib_audio.mediaplayer.events.*
import com.imooc.lib_audio.mediaplayer.model.AudioBean
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import java.lang.NullPointerException
import kotlin.random.Random

object AudioController {


    /**
     * 播放模式
     */
    enum class PlayMode{
        LOOP,
        RANDOM,
        REPEAT
    }
    
    //播放器
    private var mAudioPlayer: AudioPlayer = AudioPlayer()

    //当前播放队列
    private var mQueue: ArrayList<AudioBean> = ArrayList()

    //当前播放模式
    private lateinit var mPlayMode: PlayMode
    //当前播放索引
    private var mQueueIndex = 0

    //object 关键字表示的单例类 不允许有构造方法
    init {
        mQueueIndex = 0
        mPlayMode = PlayMode.LOOP
        EventBus.getDefault().register(this)
    }


    fun isStartState():Boolean{
        return CustomMediaPlayer.Status.STARTED == getStatus()
    }

    fun isPauseState():Boolean{
        return CustomMediaPlayer.Status.PAUSED == getStatus()
    }

    //获取当前播放队列
    fun getQueue() : ArrayList<AudioBean>{
        return mQueue
    }

    //添加Audio
    fun addAudio(bean : AudioBean){
        addAudio(0, bean)
    }

    fun addAudio(index : Int, bean : AudioBean){
        val query = mQueue.indexOf(bean)
        if(query <= -1){
            mQueue.add(index, bean)
            //设置播放索引
            setPlayIndex(index)
        }else{
            val currentPlaying = getCurrentPlaying()
            if(!currentPlaying.id.equals(bean.id)){
                setPlayIndex(index)
            }
        }
    }

    //移除音乐  不可移除当前正在播放的音乐
    fun removeAudio(bean : AudioBean){
        if(mQueue.remove(bean)){
            EventBus.getDefault().post(AudioRemoveEvent())
        }
    }

    //移除全部播放队列
    fun removeAudio(){
        val currentBean = getCurrentPlaying()
        mQueue.clear()
        //不移除当前播放的歌曲
        mQueue.add(currentBean)
    }

    //设置播放队列
    fun setQueue(bean : ArrayList<AudioBean>){
        setQueue(bean, 0)
    }

    fun setQueue(bean : ArrayList<AudioBean>, index : Int){
        mQueue.addAll(bean)
        mQueueIndex = index
    }

    //播放模式
    fun getPlayMode():PlayMode{
        return mPlayMode
    }

    //设置播放模式
    fun setPlayMode(playMode: PlayMode){
        mPlayMode = playMode
        //对外发送事件
        EventBus.getDefault().post(AudioPlayModeEvent(mPlayMode))
    }


    private fun getCurrentPlaying(): AudioBean {
        return getPlaying(mQueueIndex)
    }

    //设置播放索引
    fun setPlayIndex(index: Int) {
        mQueueIndex = index
        play()
    }

    //播放音乐
    public fun play() {
        mAudioPlayer.load(getCurrentPlaying())
    }

    //暂停播放
    fun pause(){
        mAudioPlayer.pause()
    }

    //跳转到指定时间
    fun seekTo(time : Long){
        mAudioPlayer.seekTo(time)
    }

    fun release(){
        mAudioPlayer.release()
        EventBus.getDefault().unregister(this)
    }

    fun resume(){
        mAudioPlayer.resume()
    }

    fun next(){
        mAudioPlayer.load(getNextPlaying())
    }

    //获取下一首播放的歌曲 根据当前播放模式
    private fun getNextPlaying(): AudioBean {
        when(mPlayMode){
            //列表循环
            PlayMode.LOOP -> {
                mQueueIndex = (mQueueIndex + 1) % mQueue.size
                return getPlaying(mQueueIndex)
            }
            //随机播放
            PlayMode.RANDOM -> {
                mQueueIndex = Random.nextInt(mQueue.size) % mQueue.size
                return getPlaying(mQueueIndex)
            }
            //单曲循环
            PlayMode.REPEAT -> {
                return getPlaying(mQueueIndex)
            }
        }
    }

    //获取上一首歌曲
    fun getPrePlaying():AudioBean{
        when(mPlayMode){
            //列表循环
            PlayMode.LOOP -> {
                mQueueIndex = (mQueueIndex + mQueue.size - 1 ) % mQueue.size
                return getPlaying(mQueueIndex)
            }
            //随机播放
            PlayMode.RANDOM -> {
                mQueueIndex = Random.nextInt(mQueue.size) % mQueue.size
                return getPlaying(mQueueIndex)
            }
            //单曲循环
            PlayMode.REPEAT -> {
                return getPlaying(mQueueIndex)
            }
        }
    }

    //播放上一首
    fun previous(){
        mAudioPlayer.load(getPrePlaying())
    }

    //切换 播放/暂停 状态
    fun playOrPause(){
        if(CustomMediaPlayer.Status.INITIALIZED == getStatus()){
            play()
        }
        if(isStartState()){
            pause()
        }else if(isPauseState()){
            resume()
        }
    }

    fun getNowPlaying(): AudioBean{
        return getPlaying(mQueueIndex)
    }

    //获取当前播放状态
    private fun getStatus(): CustomMediaPlayer.Status {
        return mAudioPlayer.getStatus()
    }


    fun getPlaying(index : Int): AudioBean {
        return if(!mQueue.isEmpty() && index < mQueue.size && index >= 0){
            mQueue.get(index)
        } else {
            throw NullPointerException("")
        }
    }
    //插放完毕事件处理
    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onAudioCompleteEvent(event: AudioCompleteEvent){
        //播放下一首
        next()
    }

    //播放出错事件处理
    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onAudioErrorEvent(event: AudioErrorEvent){
        //播放下一首
        next()
    }

}